### PR TITLE
feat(agent-bff): validate filters, sorts, and projections from capabilities

### PR DESCRIPTION
**Pull request by nbouliol**
## What

Adds a reusable, pure validator in `agent-bff` that checks a list/count request's `filter`, `sort`, and `projection` fields against the target collection's capabilities **before any agent call**, plus an operator normalizer. Capabilities are the only source of truth — no hardcoded type table is consulted, which fixes the drift between the frontend operator tables and what the agent actually supports.

This is the reusable validator only. Wiring it into the endpoints is out of scope (PRD-676/677).

## How

- **`operator-normalizer.ts`** — maps agent snake_case capabilities operators to the canonical PascalCase set. The snake→Pascal map is derived from datasource-toolkit `allOperators` using the same transform the agent uses to serialize them, so it round-trips every operator and cannot drift.
- **`validation-errors.ts`** — three error factories owned by this slice, reusing the existing error-envelope shape (`{ error: { type, status, message, details } }`) without touching the shared registry.
- **`capabilities-validator.ts`** — `validateAgainstCapabilities()` returns every offending field as a `BffHttpError[]` (empty = valid); `assertValidAgainstCapabilities()` throws the first. Walks every filter leaf (including under `And`/`Or` groups).

### Validation rules

| Surface | Rule | Error |
|---|---|---|
| filter | field absent from capabilities | `422 unknown_field` |
| filter | field present with no operators | `422 field_not_filterable` |
| filter | operator not in the field's normalized operators | `400 invalid_filter_operator` (+ `validOperators`) |
| sort | field absent | `422 unknown_field` |
| projection | field absent | `422 unknown_field` |

Sort/projection check existence only; operator checks apply to filters. Each error carries the offending `details.field`.

## Notable decisions

- **Normalize against the full `allOperators` set**, not just `uniqueOperators` as the ticket wording suggests — real capabilities include `in`, `contains`, `blank`, etc.
- **A relation field appears in capabilities with no `operators` key** (not `operators: []`); treated as `field_not_filterable` via `operators ?? []`. `CapabilitiesResult.operators` made optional to match runtime. See the ticket comment for the invariant discrepancy.
- **An unmapped capabilities operator throws `500 mapping_error`** (loud) rather than dropping or passing it through — this only happens on an agent/BFF version skew.

## Tests

22 new focused unit tests: normalizer round-trip over all operators + edge cases; every error case; a fully-valid request; a mapping_error case; and an assertion that a field absent from capabilities is rejected with no type table consulted. Full suite: 660/660 green.

Fixes PRD-675

🤖 Generated with [Claude Code](https://claude.com/claude-code)

> [!NOTE]
> ### Validate filters, sorts, and projections against agent capabilities in data routes
> - Adds local validation of filter operators, sort fields, and projection fields against fetched agent capabilities before forwarding list and count requests, returning structured 4xx errors without calling the agent when validation fails.
> - Introduces [`capabilities-validator.ts`](https://github.com/ForestAdmin/agent-nodejs/pull/1755/files#diff-cff066ad8696ed33f89763a6029006dcc2e1dda09c911f5c6630a75b2ae1c3b3) with full filter AST traversal (capped at `MAX_FILTER_DEPTH`), operator normalization via [`operator-normalizer.ts`](https://github.com/ForestAdmin/agent-nodejs/pull/1755/files#diff-653e0cee258f214240011bcbb55314be0494a4c2965438d6b4abbfea94e51f74), and structured error factories in [`validation-errors.ts`](https://github.com/ForestAdmin/agent-nodejs/pull/1755/files#diff-6e6cf3f081c14b4bfd6388ceff252fadf9ef08bb2945b665277858fa3e5cee45).
> - Capabilities are fetched only when the request carries filter/sort/projection, controlled by the new `hasCapabilityConstrainedInput` predicate.
> - Fixes `action-execute-mapper` to drop non-string entries from `refresh.relationships` and reject malformed webhook payloads.
> - Behavioral Change: list and count requests with unsupported operators, unknown fields, or filters exceeding max depth now fail locally with structured errors instead of being forwarded to the agent.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 7870073.</sup>
> 
>

### CODE REVIEW

File: packages/agent-bff/src/validation/capabilities-validator.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    if (operators === undefined) {
+      errors.push(unknownField(leaf.field));
+    } else if (operators.length === 0) {
+      errors.push(fieldNotFilterable(leaf.field));
+    } else if (leaf.operator !== undefined) {
+      const normalized = normalizeFieldOperators(leaf.field, operators);
+
+      if (!normalized.includes(leaf.operator)) {
+        errors.push(invalidFilterOperator(leaf.field, normalized));
+      }
```
[the remark this answers is not in the document]
Comment (nbouliol) [reply packages/agent-bff/src/validation/capabilities-validator.ts]: Fixed in dc02e94c9. A condition-tree leaf without a valid operator is now rejected with `invalid_filter_operator` (listing `validOperators`) before any agent call. Covered by two new tests: a leaf with no operator, and a leaf with a non-string operator.
